File: .NET/vb.net/Discoverer.vb
Friend Delegate Sub DeviceEvent(sender As Object, Address As Int64)

Friend Class Discoverer
    Inherits wclBluetoothManager

    Private FFoundDevices As List(Of Int64)
    Private FDevices As List(Of Int64)
    Private FDiscovering As Boolean
    Private FRadio As wclBluetoothRadio

    Protected Overrides Sub DoDiscoveringStarted(Radio As wclBluetoothRadio)
        FFoundDevices.Clear()
    End Sub

    Protected Overrides Sub DoDeviceFound(Radio As wclBluetoothRadio, Address As Int64)
        FFoundDevices.Add(Address)
        If Not FDevices.Contains(Address) Then
            FDevices.Add(Address)
            If OnDeviceFoundEvent IsNot Nothing Then OnDeviceFoundEvent(Me, Address)
        End If
    End Sub

    Protected Overrides Sub DoDiscoveringCompleted(Radio As wclBluetoothRadio, [Error] As Integer)
        Dim Devices As List(Of Int64) = New List(Of Int64)()

        For Each Address As Int64 In FDevices
            If FFoundDevices.Contains(Address) Then
                Devices.Add(Address)
            Else
                If OnDeviceLostEvent IsNot Nothing Then OnDeviceLostEvent(Me, Address)
            End If
        Next

        FDevices.Clear()
        FDevices = Devices

        If FDiscovering Then Radio.Discover(10, wclBluetoothDiscoverKind.dkClassic)
    End Sub

    Sub New()
        MyBase.New()

        FFoundDevices = Nothing
        FDevices = Nothing
        FDiscovering = False
        FRadio = Nothing

        OnDeviceFoundEvent = Nothing
        OnDeviceLostEvent = Nothing
        OnStartedEvent = Nothing
        OnStoppedEvent = Nothing
    End Sub

    Public Function Start() As Integer
        If FDiscovering Then Return wclBluetoothErrors.WCL_E_BLUETOOTH_DISCOVERING_RUNNING
        If Count = 0 Then Return wclBluetoothErrors.WCL_E_BLUETOOTH_HARDWARE_NOT_AVAILABLE

        Dim Radio As wclBluetoothRadio = Nothing
        For i As Integer = 0 To Count - 1
            If Me(i).Available Then
                Radio = Me(i)
                Exit For
            End If
        Next

        If Radio Is Nothing Then Return wclBluetoothErrors.WCL_E_BLUETOOTH_DRIVER_NOT_AVAILABLE

        FDevices = New List(Of Int64)()
        FFoundDevices = New List(Of Int64)()

        Dim Res As Integer = Radio.Discover(10, wclBluetoothDiscoverKind.dkClassic)
        If Res <> wclErrors.WCL_E_SUCCESS Then
            FDevices = Nothing
            FFoundDevices = Nothing
        Else
            FDiscovering = True
            FRadio = Radio

            If OnStartedEvent IsNot Nothing Then OnStartedEvent(Me, EventArgs.Empty)
        End If

        Return Res
    End Function

    Public Function [Stop]() As Integer
        If Not FDiscovering Then Return wclBluetoothErrors.WCL_E_BLUETOOTH_DISCOVERING_NOT_RUNNING

        FDiscovering = False
        Dim Res As Integer = FRadio.Terminate()

        FDevices.Clear()
        FFoundDevices.Clear()

        FRadio = Nothing
        FDevices = Nothing
        FFoundDevices = Nothing

        If OnStoppedEvent IsNot Nothing Then OnStoppedEvent(Me, EventArgs.Empty)

        Return Res
    End Function

    Public ReadOnly Property Radio As wclBluetoothRadio
        Get
            Return FRadio
        End Get
    End Property

    Public Shadows Event OnDeviceFound As DeviceEvent
    Public Event OnDeviceLost As DeviceEvent
    Public Event OnStarted As EventHandler
    Public Event OnStopped As EventHandler
End Class

File: .NET/c#/Discoverer.cs
using System;
using System.Collections.Generic;

using wclCommon;
using wclBluetooth;

namespace ContinueDiscovering
{
    internal delegate void DeviceEvent(Object sender, Int64 Address);

    internal class Discoverer : wclBluetoothManager
    {
        private List<Int64> FFoundDevices;
        private List<Int64> FDevices;
        private Boolean FDiscovering;
        private wclBluetoothRadio FRadio;

        protected override void DoDiscoveringStarted(wclBluetoothRadio Radio)
        {
            FFoundDevices.Clear();
        }

        protected override void DoDeviceFound(wclBluetoothRadio Radio, Int64 Address)
        {
            FFoundDevices.Add(Address);
            if (!FDevices.Contains(Address))
            {
                FDevices.Add(Address);
                if (OnDeviceFound != null)
                    OnDeviceFound(this, Address);
            }
        }

        protected override void DoDiscoveringCompleted(wclBluetoothRadio Radio, int Error)
        {
            List<Int64> Devices = new List<Int64>();

            foreach (Int64 Address in FDevices)
            {
                if (FFoundDevices.Contains(Address))
                    Devices.Add(Address);
                else
                {
                    if (OnDeviceLost != null)
                        OnDeviceLost(this, Address);
                }
            }

            FDevices.Clear();
            FDevices = Devices;

            if (FDiscovering)
                Radio.Discover(10, wclBluetoothDiscoverKind.dkClassic);
        }

        public Discoverer()
            : base()
        {
            FFoundDevices = null;
            FDevices = null;
            FDiscovering = false;
            FRadio = null;

            OnDeviceFound = null;
            OnDeviceLost = null;
            OnStarted = null;
            OnStopped = null;
        }

        public Int32 Start()
        {
            if (FDiscovering)
                return wclBluetoothErrors.WCL_E_BLUETOOTH_DISCOVERING_RUNNING;

            if (Count == 0)
                return wclBluetoothErrors.WCL_E_BLUETOOTH_HARDWARE_NOT_AVAILABLE;

            wclBluetoothRadio Radio = null;
            for (int i = 0; i < Count; i++)
            {
                if (this[i].Available)
                {
                    Radio = this[i];
                    break;
                }
            }

            if (Radio == null)
                return wclBluetoothErrors.WCL_E_BLUETOOTH_DRIVER_NOT_AVAILABLE;

            FDevices = new List<Int64>();
            FFoundDevices = new List<Int64>();

            Int32 Res = Radio.Discover(10, wclBluetoothDiscoverKind.dkClassic);
            if (Res != wclErrors.WCL_E_SUCCESS)
            {
                FDevices = null;
                FFoundDevices = null;
            }
            else
            {
                FDiscovering = true;
                FRadio = Radio;

                if (OnStarted != null)
                    OnStarted(this, EventArgs.Empty);
            }

            return Res;
        }

        public Int32 Stop()
        {
            if (!FDiscovering)
                return wclBluetoothErrors.WCL_E_BLUETOOTH_DISCOVERING_NOT_RUNNING;

            FDiscovering = false;
            int Res = FRadio.Terminate();

            FDevices.Clear();
            FFoundDevices.Clear();

            FRadio = null;
            FDevices = null;
            FFoundDevices = null;

            if (OnStopped != null)
                OnStopped(this, EventArgs.Empty);

            return Res;
        }

        public wclBluetoothRadio Radio { get { return FRadio; } }

        public new event DeviceEvent OnDeviceFound;
        public event DeviceEvent OnDeviceLost;
        public event EventHandler OnStarted;
        public event EventHandler OnStopped;
    };
}
